server/service/endpoint_utils_test.go

```diff
@@ -240,3 +249,149 @@ func TestUniversalDecoderQueryAndListPlayNice(t *testing.T) {
 	require.NotNil(t, casted.ID1)
 	assert.Equal(t, uint(444), *casted.ID1)
 }
+
+func TestEndpointer(t *testing.T) {
+
+	r := mux.NewRouter()
+	ds := new(mock.Store)
+	ds.SessionByKeyFunc = func(ctx context.Context, key string) (*fleet.Session, error) {
+		return &fleet.Session{
+			ID:         3,
+			UserID:     42,
+			Key:        key,
+			AccessedAt: time.Now(),
+		}, nil
+	}
+	ds.DestroySessionFunc = func(ctx context.Context, session *fleet.Session) error {
+		return nil
+	}
+	ds.MarkSessionAccessedFunc = func(ctx context.Context, session *fleet.Session) error {
+		return nil
+	}
+	ds.UserByIDFunc = func(ctx context.Context, id uint) (*fleet.User, error) {
+		return &fleet.User{GlobalRole: ptr.String(fleet.RoleAdmin)}, nil
+	}
+	ds.ListUsersFunc = func(ctx context.Context, opt fleet.UserListOptions) ([]*fleet.User, error) {
+		return []*fleet.User{{GlobalRole: ptr.String(fleet.RoleAdmin)}}, nil
+	}
+
+	svc := newTestService(ds, nil, nil)
+
+	fleetAPIOptions := []kithttp.ServerOption{
+		kithttp.ServerBefore(
+			kithttp.PopulateRequestContext, // populate the request context with common fields
+			setRequestsContexts(svc),
+		),
+		kithttp.ServerErrorHandler(&errorHandler{kitlog.NewNopLogger()}),
+		kithttp.ServerErrorEncoder(encodeError),
+		kithttp.ServerAfter(
+			kithttp.SetContentType("application/json; charset=utf-8"),
+			logRequestEnd(kitlog.NewNopLogger()),
+			checkLicenseExpiration(svc),
+		),
+	}
+
+	e := NewUserAuthenticatedEndpointer(svc, fleetAPIOptions, r, "v1", "2021-11")
+	nopHandler := func(ctx context.Context, request interface{}, svc fleet.Service) (interface{}, error) {
+		if authctx, ok := authz_ctx.FromContext(ctx); ok {
+			authctx.SetChecked()
+		}
+		return "nop", nil
+	}
+	overrideHandler := func(ctx context.Context, request interface{}, svc fleet.Service) (interface{}, error) {
+		if authctx, ok := authz_ctx.FromContext(ctx); ok {
+			authctx.SetChecked()
+		}
+		return "override", nil
+	}
+
+	// Regular path, no plan to deprecate
+	e.GET("/api/v1/fleet/path1", nopHandler, struct{}{})
+
+	// New path, we want it only available starting from the specified version
+	e.StartingAtVersion("2021-11").GET("/api/v1/fleet/newpath", nopHandler, struct{}{})
+
+	// Path that was in v1, but was changed in 2021-11
+	e.EndingAtVersion("v1").GET("/api/v1/fleet/overriddenpath", nopHandler, struct{}{})
+	e.StartingAtVersion("2021-11").GET("/api/v1/fleet/overriddenpath", overrideHandler, struct{}{})
+
+	// Path that got deprecated
+	e.EndingAtVersion("v1").GET("/api/v1/fleet/deprecated", nopHandler, struct{}{})
+	// Path that got deprecated but in the latest version
+	e.EndingAtVersion("2021-11").GET("/api/v1/fleet/deprecated-soon", nopHandler, struct{}{})
+
+	// Aliasing works with versioning too
+	e.WithAlias("/api/v1/fleet/something/{fff}").GET("/api/v1/fleet/somethings/{fff}", nopHandler, struct{}{})
```

Comment:
Aahh I thought the alias feature was for versions, i.e. `v1` is an alias of `2021-11`. I think something like `WithAltPath` could be a bit clearer? I think even better would be to allow passing multiple paths to `e.GET`, `e.POST`, etc. but this is a bigger change, so I get the appeal of going for an extra method for this, was just assuming the wrong thing with the name.